In [2]:
[i for i in range(10)]

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

- List Comprehension. Cum am putea folosi generators ? Very EASY -> folosim ()

In [3]:
(i for i in range(10))

<generator object <genexpr> at 0x7a2f841413c0>

- Generatorul nu salveaza lista in memorie, de asta nu returneaza [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [4]:
import sys

In [5]:
a = [i for i in range(100)]
b = (i for i in range(100))

In [6]:
sys.getsizeof(a)

920

In [7]:
sys.getsizeof(b)

192

In [8]:
a = [i for i in range(10000)]
b = (i for i in range(10000))

In [9]:
print(sys.getsizeof(a))
print(sys.getsizeof(b))

85176
192


-   Generatorul ramane la 192, chiar daca crestem range-ul x100. Pentru ca nu salveaza datele in memorie, se poate vedea cum ocupa semnificativ mai putin

In [10]:
from timeit import timeit

In [11]:
timeit("(i for i in range(100))")

0.08820302800450008

In [12]:
timeit("[i for i in range(100)]")

0.6424888770052348

-   Generators -> Small size in memory and increased time performance

In [13]:
list(b)[0]

0